### PR TITLE
Camera clean manip

### PR DESCRIPTION
**Pull request by dlyr**
Missing clean of Camera
+ improve trackball.

* **Please check if the PR fulfills these requirements**
- [x] The commit message follows our guidelines
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

* **What kind of change does this PR introduce?** (Bug fix, feature, docs update, ...)
Well .. see the commits ;)
Basically a new way to handle trackball, and also some fixes for camera, cleanup.

* **What is the current behavior?** (You can also link to an open issue here)
Trackball do not work for arbitrary camera ...

* **What is the new behavior (if this is a feature change)?**
now it seems to work.

* **Does this PR introduce a breaking change?** (What changes might users need to make in their application due to this PR?)
Not so much.

### CODE REVIEW

File: src/Gui/Viewer/TrackballCameraManipulator.cpp [hunk 1]
```diff
@@ -167,11 +139,14 @@ bool Gui::TrackballCameraManipulator::handleMouseMoveEvent(
     else
     { m_quickCameraModifier = 2.0_ra; }
 
-    if ( m_cameraRotateMode ) { handleCameraRotate( dx, dy ); }
-
-    if ( m_cameraPanMode ) { handleCameraPan( dx, dy ); }
-
-    if ( m_cameraZoomMode ) { handleCameraZoom( dx, dy ); }
+    if ( m_currentAction == TRACKBALLCAMERA_ROTATE )
```
Comment (nmellado) [inline src/Gui/Viewer/TrackballCameraManipulator.cpp:142]: Use switch case ?

File: src/Gui/Viewer/TrackballCameraManipulator.cpp [hunk 2]
```diff
@@ -184,22 +159,25 @@ bool Gui::TrackballCameraManipulator::handleMouseMoveEvent(
...
-bool Gui::TrackballCameraManipulator::handleWheelEvent( QWheelEvent* event ) {
+bool Gui::TrackballCameraManipulator::handleWheelEvent( QWheelEvent* event,
+                                                        const Qt::MouseButtons& buttons,
+                                                        const Qt::KeyboardModifiers& modifiers,
+                                                        int key
 
-    handleCameraZoom( ( event->angleDelta().y() * 0.01_ra + event->angleDelta().x() * 0.01_ra ) *
-                      m_wheelSpeedModifier );
+) {
+    ///\todo use action
```
Comment (nmellado) [inline src/Gui/Viewer/TrackballCameraManipulator.cpp:177]: I do not understand what this todo means

Comment (dlyr) [reply src/Gui/Viewer/TrackballCameraManipulator.cpp:177]: will be done ;)

**Review by MathiasPaulin — CHANGES_REQUESTED** [review]
I don't understand why PR #740 was approved and merged so quickly. This one seems to bring a lot of updates but still need to be validated on several uses cases.

**Comment by dlyr** [comment]
Well the main guideline is, I think, there is not regression... but his one will improve and polish what have been done. (and one month is not so quick)

File: src/Gui/Viewer/FlightCameraManipulator.cpp [hunk 3]
```diff
@@ -195,22 +182,31 @@ bool Gui::FlightCameraManipulator::handleMouseMoveEvent( QMouseEvent* event,
 
     emit cameraChanged( m_camera->getPosition(), m_target );
 
-    return m_cameraRotateMode || m_cameraPanMode || m_cameraZoomMode;
+    return m_currentAction.isValid();
```
[the remark this answers is not in the document]
Comment (dlyr) [reply src/Gui/Viewer/FlightCameraManipulator.cpp:185]: yes (it was the case before)

File: src/Gui/Viewer/CameraManipulator.hpp [hunk 4]
```diff
@@ -149,23 +140,26 @@ class RA_GUI_API CameraManipulator : public QObject
...
-    Core::Vector3 m_target;
+    Core::Vector3 m_target {0_ra, 0_ra, 0_ra};
+
+    /// Whether the corresponding camera movement is active or not.
+    KeyMappingManager::KeyMappingAction m_currentAction {};
 
-    Core::Asset::Camera* m_camera; ///< The Camera.
-    Engine::Scene::Light* m_light; ///< The light attached to the Camera.
+    Core::Asset::Camera* m_camera {nullptr}; ///< The Camera.
+    Engine::Scene::Light* m_light {nullptr}; ///< The light attached to the Camera.
```
[the remark this answers is not in the document]
Comment (dlyr) [reply src/Gui/Viewer/CameraManipulator.hpp:162]: don't know, out of scope of the PR.

File: src/Gui/Viewer/TrackballCameraManipulator.cpp [hunk 5]
```diff
@@ -184,46 +159,56 @@ bool Gui::TrackballCameraManipulator::handleMouseMoveEvent(
 
     emit cameraChanged( m_camera->getPosition(), m_target );
```
[the remark this answers is not in the document]
Comment (dlyr) [reply src/Gui/Viewer/TrackballCameraManipulator.cpp:160]: out of scope ?

[the remark this answers is not in the document]
Comment (dlyr) [reply src/Gui/Viewer/TrackballCameraManipulator.cpp:160]: well previous state seems to double emit cameraChanged (one time in handle sub method, one time in the main one, I'll double check)

File: src/Gui/Viewer/TrackballCameraManipulator.cpp [hunk 6]
```diff
@@ -184,46 +159,56 @@ bool Gui::TrackballCameraManipulator::handleMouseMoveEvent(
...
+    using ProjType = Ra::Core::Asset::Camera::ProjType;
+    if ( action == TRACKBALLCAMERA_PROJ_MODE )
     {
-        m_rotateAround = !m_rotateAround;
-        return true;
+        m_camera->setType( m_camera->getType() == ProjType::ORTHOGRAPHIC ? ProjType::PERSPECTIVE
+                                                                         : ProjType::ORTHOGRAPHIC );
     }
 
     return false;
```
[the remark this answers is not in the document]
Comment (dlyr) [reply src/Gui/Viewer/TrackballCameraManipulator.cpp:214]: right.

File: src/Gui/Viewer/Viewer.cpp [hunk 7]
```diff
@@ -592,7 +582,7 @@ void Viewer::showEvent( QShowEvent* ev ) {
     WindowQt::showEvent( ev );
     /// \todo remove this commented code when camera init in ctr is tested on other arch.
```
[the remark this answers is not in the document]
Comment (dlyr) [reply src/Gui/Viewer/Viewer.cpp:583]: good question

File: tests/ExampleApps/CustomCameraManipulator/main.cpp [hunk 8]
```diff
@@ -28,25 +28,21 @@ class CameraManipulator2D : public Ra::Gui::TrackballCameraManipulator
...
-        return handled;
+        m_currentAction = Ra::Gui::KeyMappingManager::getInstance()->getAction(
+            Ra::Gui::TrackballCameraManipulator::TrackballCameraMapping::getContext(),
+            buttons,
+            modifiers,
+            key,
+            false );
+
+        // ignore rotate
+        if ( m_currentAction == TRACKBALLCAMERA_ROTATE )
```
[the remark this answers is not in the document]
Comment (dlyr) [reply tests/ExampleApps/CustomCameraManipulator/main.cpp:42]: what do you mean ? which hacking ? 
I do not see what is missing here ...

File: src/Gui/Viewer/FlightCameraManipulator.cpp [hunk 9]
```diff
@@ -220,7 +216,7 @@ bool Gui::FlightCameraManipulator::handleWheelEvent( QWheelEvent* event ) {
 
     emit cameraPositionChanged( m_camera->getPosition() );
 
-    return true;
+    return action.isValid();
```
[the remark this answers is not in the document]
Comment (dlyr) [reply src/Gui/Viewer/FlightCameraManipulator.cpp:219]: not sure to see which question

File: src/Gui/Configs/default.xml [hunk 10]
```diff
@@ -10,13 +10,20 @@
             action="TRACKBALLCAMERA_PAN"/>
     <keymap context="CameraContext" key="" modifiers="ControlModifier" buttons="MiddleButton"
             action="TRACKBALLCAMERA_ZOOM"/>
-    <keymap context="CameraContext" key="Key_A" modifiers="" buttons="" action="TRACKBALLCAMERA_ROTATE_AROUND"/>
+    <keymap context="CameraContext" key="" modifiers="AltModifier,ControlModifier" buttons="MiddleButton"
```
[the remark this answers is not in the document]
Comment (dlyr) [reply src/Gui/Configs/default.xml:13]: sure, let's mac and alternative user do it.

File: src/Gui/Viewer/FlightCameraManipulator.cpp [hunk 11]
```diff
@@ -195,22 +182,31 @@ bool Gui::FlightCameraManipulator::handleMouseMoveEvent( QMouseEvent* event,
 
     emit cameraChanged( m_camera->getPosition(), m_target );
```
[the remark this answers is not in the document]
Comment (dlyr) [reply src/Gui/Viewer/FlightCameraManipulator.cpp:183]: I temporary remove all signals, I do not see any usage of that

File: src/Gui/Viewer/FlightCameraManipulator.cpp [hunk 12]
```diff
@@ -195,22 +182,31 @@ bool Gui::FlightCameraManipulator::handleMouseMoveEvent( QMouseEvent* event,
...
 
     return true;
 }
 
-bool Gui::FlightCameraManipulator::handleWheelEvent( QWheelEvent* event ) {
+bool Gui::FlightCameraManipulator::handleWheelEvent( QWheelEvent* event,
+                                                     const Qt::MouseButtons& buttons,
+                                                     const Qt::KeyboardModifiers& modifiers,
+                                                     int key ) {
+    ///\todo use action.
```
[the remark this answers is not in the document]
Comment (dlyr) [reply src/Gui/Viewer/FlightCameraManipulator.cpp:199]: done.

**Comment by dlyr** [comment]
Well I have an issue with TrackballManipulator (which is not linked to the current PR I think)
It seems that phi theta computation are totally *** in many cases ...
But it takes time to debug, and we have a new manipulator incoming.

So I propose to simplify trackball as a "real theta phi rotate manipulator in world space around a target"
which can be done in the current PR.

File: src/Core/Asset/Camera.hpp [hunk 13]
```diff
@@ -96,6 +103,9 @@ class RA_CORE_API Camera
     inline Scalar getFOV() const;
 
     /// Set the Field Of View to 'fov' in the x (horizontal) direction.
+    /// If you have an vertical field of view, you can convert it to horizontal as
```
Comment (MathiasPaulin) [inline src/Core/Asset/Camera.hpp:106]:
```suggestion
    /// If you have a vertical field of view, you can convert it to horizontal as
```

File: src/Engine/Scene/CameraComponent.hpp [hunk 14]
```diff
@@ -13,6 +13,7 @@ namespace Scene {
 class Entity;
 
 /// A Camera is an Engine Component storing a Camera object.
+///\todo attached camera do not take entity's transformation into account ...
```
Comment (MathiasPaulin) [inline src/Engine/Scene/CameraComponent.hpp:16]: Do this means that if the Entity is animated (e.g. keyframed) there is no impact on the Camera ?

File: src/Engine/Scene/CameraComponent.inl [hunk 15]
```diff
@@ -5,115 +5,6 @@
...
-
-inline Core::Vector2 CameraComponent::project( const Core::Vector3& p ) const {
-    return m_camera->project( p ).head<2>();
-}
-
-inline Core::Vector3 CameraComponent::unProject( const Core::Vector2& pix ) const {
-    return m_camera->unProject( pix );
-}
-} // namespace Scene
+namespace Scene {} // namespace Scene
```
[the remark this answers is not in the document]
Comment (MathiasPaulin) [reply src/Engine/Scene/CameraComponent.inl:8]: The question was marked resolved without answer nor removal of this empty namespace declaration. What's this declaration aim ?

File: src/Gui/Viewer/TrackballCameraManipulator.cpp [hunk 16]
```diff
@@ -347,68 +328,86 @@ void Gui::TrackballCameraManipulator::handleCameraPan( Scalar dx, Scalar dy ) {
...
 
-    m_phi = ( areApproxEqual( R.z(), 0_ra ) && areApproxEqual( R.x(), 0_ra ) )
-                ? std::acos( m_camera->getRightVector().dot( Ra::Core::Vector3::UnitZ() ) ) + Pi
-                : std::atan2( R.z(), R.x() );
+    m_theta = std::acos( R.y() );
 
-    // Keep phi between 0 and 2pi
-    // Keep theta between -pi and pi
-    if ( m_phi < 0_ra )
+    // unlikly to have z and x to 0, unless direction is perfectly aligne with m_referenceFrame.z()
```
Comment (MathiasPaulin) [inline src/Gui/Viewer/TrackballCameraManipulator.cpp:370]:
```suggestion
    // unlikely to have z and x to 0, unless direction is perfectly aligned with m_referenceFrame.z()
```

File: src/Gui/Viewer/Viewer.hpp [hunk 17]
```diff
@@ -209,18 +209,21 @@ class RA_GUI_API Viewer : public WindowQt, public KeyMappingManageable<Viewer>
     getPickingMode( const Ra::Gui::KeyMappingManager::KeyMappingAction& action ) const;
 
     /// @name
-    /// Qt event, do the viewer stuff, and call handle*Event to perform the
-    /// actual event handling, according to keyMapping.
+    /// If GL is initialized, Do the viewer stuff, and call handle*Event to perform actual event
```
Comment (MathiasPaulin) [inline src/Gui/Viewer/Viewer.hpp:212]:
```suggestion
    /// If GL is initialized, do the viewer stuff, and call handle*Event to perform actual event
```

**Review by MathiasPaulin — COMMENTED** [review]
Seems to converge toward an accept. I'll test it on several configurations and, if OK, I'll approve. BTW, the few remarks below should be taken into account.

File: src/Engine/Scene/CameraComponent.hpp (hunk 14 above)
Comment (dlyr) [reply src/Engine/Scene/CameraComponent.hpp:16]: yes, but if the camera is animated, the component's RO is updated.

Comment (MathiasPaulin) [inline src/Engine/Scene/CameraComponent.hpp:16]:
```suggestion
///\todo attached camera do not take entity's transformation into account. This prevents, for now, keyframe animation of the Camera from its entity ...
```

Comment (dlyr) [reply src/Engine/Scene/CameraComponent.hpp:16]: @MathiasPaulin actually if you have a simple example of how one can imagine animate a camera's entity, I'll can do a quick fix.

**Review by MathiasPaulin — APPROVED** [review]

**Comment by dlyr** [comment]
Fine on my side.
After a quick look I do not find a easy way to have camera/component/entity transformation all synchronize ... let's open an issue

**Review by MathiasPaulin — APPROVED** [review]

**Review by MathiasPaulin — CHANGES_REQUESTED** [review]
Seems there is a problem with key-mapping configuration of the manipulator on macOS. Need to test and check.

**Review by MathiasPaulin — DISMISSED** [review]

**Comment by MathiasPaulin** [comment]
The problem I mentioned before seems to appear randomly. Investigation will be done after this PR is merged.

**Review by nmellado — APPROVED** [review]

**Pull request merged**